37-android-improve-test-coverage-of-language-screen-composable

viamgr opened the pull request:
Wrote some unit tests to improve the coverage of package com.epiclabs.thinktok.main.ui.languageselection package

viamgr committed:
Add instrumentation tests and test tags for LanguageDropdown

- Added UI tests for `LanguageDropdown` to verify initial state, expansion, selection, and display of all languages.
- Added test tags to `LanguageDropdown` components to make UI testing easier.
- Added `debugImplementation` and `androidTestImplementation` for `androidx.ui.test.manifest` to `build.gradle.kts` for testing purposes.
- Added `androidTestImplementation` for `androidx.ui.test.junit4` to `build.gradle.kts` for testing purposes.

viamgr committed:
Refactored tests and added Robolectric support

- Moved `LanguageDropdownTest` from `androidTest` to `test` source set.
- Added Robolectric test runner to `LanguageDropdownTest`.
- Added Robolectric dependency in `libs.versions.toml`.
- Enabled including Android resources for unit tests in both `AppModulePlugin` and `AndroidModulePlugin`.
- Removed unnecessary `testImplementation` of `androidx.ui.test.manifest` from feature module and added it `testImplementation` instead of `androidTestImplementation`.

viamgr committed:
Updated dependency versions:
- Removed Kotest dependencies.

mmbehroozfar commented on feature/main/ui/src/main/kotlin/com/epiclabs/thinktok/main/ui/languageselection/LanguageDropdown.kt:
```diff
@@ -43,18 +44,19 @@ internal fun LanguageDropdown(
 
         ExposedDropdownMenu(
-            modifier = Modifier.fillMaxWidth(),
+            modifier = Modifier.fillMaxWidth().testTag("LanguageDropdownMenu"),
```
I think it's better to use a variable instead of hard-coding the tag.

Also, I have a suggestion, I think it's better not to use test tags. We can use Compose semantics instead.

What do you think?

viamgr replied:
I defined some private keys but I didn't change the tests. I know the advantages of using compose semantics, we don't need to define and maintain test keys but it brings some complexity. Both approaches have pros and cons, but we follow the easiest way and can improve it if needed.

mmbehroozfar replied:
Yes, we are on the same page!

viamgr committed:
Refactor: Added test tags to LanguageDropdown component

- Added test tags to the LanguageDropdown composable, improving UI testability.
- Introduced constants for test tag strings.
- Added a test tag to the LanguageDropdownTextField
- Added a test tag to the LanguageDropdownMenu
- Added a test tag to each LanguageDropdownMenuItem, with a dynamic suffix for the language.

Homa-Shafiei commented on feature/main/ui/src/test/java/com/epiclabs/thinktok/main/ui/languageselection/LanguageDropdownTest.kt:
```diff
@@ -0,0 +1,90 @@
+
+    @Test
+    fun languageDropdown_initialState() {
+        val languages = listOf("English", "Spanish", "French")
```
This is a duplicated variable and we can define it global

Homa-Shafiei commented on feature/main/ui/src/test/java/com/epiclabs/thinktok/main/ui/languageselection/LanguageDropdownTest.kt:
```diff
@@ -0,0 +1,90 @@
+    val composeTestRule = createComposeRule()
+
+    @Test
+    fun languageDropdown_initialState() {
```
I think we need a convention for test naming and should follow the Given-When-Then pattern.

viamgr committed:
Refactored LanguageDropdownTest test cases

- Renamed test functions to `Given-When-Then` convention for better readability.
- Extracted `languages` and `selectedLanguage` as class properties.
